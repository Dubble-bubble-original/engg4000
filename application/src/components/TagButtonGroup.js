// React
import { useState } from 'react';
import { ToggleButtonGroup, ToggleButton, InputGroup, FormControl } from 'react-bootstrap';
import PropTypes from 'prop-types'

const tagList = [
  // Location
  'forest',
  'mountain',
  'desert',
  'river',
  'lake',
  'ocean',
  'cave',
  'trail',
  'farm',
  'city',
  'building',
  'wilderness',

  // Activity
  'hiking',
  'climbing',
  'hunting',
  'walking',
  'running',
  'cycling',
  'sports',
  'tourism',
  'event',
  'food',
  'entertainment',

  // Season
  'summer',
  'fall',
  'winter',
  'spring',

  // Object
  'nature',
  'scenery',
  'sky',
  'animal',
  'art',
  'architecture',

  // Adjective
  'unique',
  'free',
  'cheap',
  'expensive',
  'underground',
  'seasonal',
  'dangerous',
  'other',
];

function TagButtonGroup(props) {
  // State variables
  const [tagFilter, setTagFilter] = useState('');

  return (
    <>
      <InputGroup className="mb-3" style={{maxWidth:250}} data-testid="filter-input">
        <InputGroup.Text>Filter:</InputGroup.Text>
        <FormControl
          type="text"
          value={tagFilter}
          id="filter-input-field"
          onChange={(e)=> {setTagFilter(e.target.value)}}
        />
      </InputGroup>
      <ToggleButtonGroup
          className="tag-container justify-content-center"
          type="checkbox"
          value={props.tags}
          onChange={(val)=>{props.setTags(val)}}
          data-testid="tag-group"
      >
        {
          tagList
            // Use regular expression to do case-insensitive matching for the filter
            .filter(tag => tagFilter ? (new RegExp(tagFilter, 'i')).test(tag) : true)
            .map(tag =>
              <ToggleButton
                variant="outline-primary"
                className="tag"
                value={tag}
                key={tag}
                id={'tgb-btn-'+tag}
                data-testid="tags"
              >
                {tag}
              </ToggleButton>
            )
        }
      </ToggleButtonGroup>
    </>
  );
}

TagButtonGroup.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.string),
  setTags: PropTypes.func.isRequired
}

export default TagButtonGroup;