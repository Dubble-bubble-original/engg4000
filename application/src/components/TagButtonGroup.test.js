import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import TagButtonGroup from './TagButtonGroup';


// Clear everything after each test
afterEach(cleanup);

// Mock functions
const setTags = jest.fn();
const tags = ['river', 'running', 'fall', 'scenery', 'seasonal'];

describe('TagButtonGroup Component', () => {

  beforeEach(() => {
    render(<TagButtonGroup tags={tags} setTags={setTags} />);
  });
    
  test('render TagButtonGroup', () => {
    expect(screen.getByTestId('filter-input')).toBeInTheDocument();
    expect(screen.getByTestId('tag-group')).toBeInTheDocument();
  });

  test('check to make sure provided tags are selected', async () => {
    for (let index = 0; index < tags.length; index++) {
      let tag = await screen.findByDisplayValue(tags[index]);
      expect(tag).toHaveProperty('checked');
      expect(tag).toMatchSnapshot();
    }
  });

  test('Sanpshot test for tagList', async () => {
    expect(screen.getAllByTestId('tags')).toMatchSnapshot();
  });
});
