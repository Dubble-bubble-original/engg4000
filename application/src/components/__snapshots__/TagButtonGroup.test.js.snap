// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagButtonGroup Component Sanpshot test for tagList 1`] = `
Array [
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-forest"
    role="button"
    tabindex="0"
  >
    forest
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-mountain"
    role="button"
    tabindex="0"
  >
    mountain
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-desert"
    role="button"
    tabindex="0"
  >
    desert
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-river"
    role="button"
    tabindex="0"
  >
    river
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-lake"
    role="button"
    tabindex="0"
  >
    lake
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-ocean"
    role="button"
    tabindex="0"
  >
    ocean
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-cave"
    role="button"
    tabindex="0"
  >
    cave
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-trail"
    role="button"
    tabindex="0"
  >
    trail
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-farm"
    role="button"
    tabindex="0"
  >
    farm
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-city"
    role="button"
    tabindex="0"
  >
    city
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-building"
    role="button"
    tabindex="0"
  >
    building
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-wilderness"
    role="button"
    tabindex="0"
  >
    wilderness
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-hiking"
    role="button"
    tabindex="0"
  >
    hiking
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-climbing"
    role="button"
    tabindex="0"
  >
    climbing
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-hunting"
    role="button"
    tabindex="0"
  >
    hunting
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-walking"
    role="button"
    tabindex="0"
  >
    walking
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-running"
    role="button"
    tabindex="0"
  >
    running
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-cycling"
    role="button"
    tabindex="0"
  >
    cycling
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-sports"
    role="button"
    tabindex="0"
  >
    sports
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-tourism"
    role="button"
    tabindex="0"
  >
    tourism
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-event"
    role="button"
    tabindex="0"
  >
    event
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-food"
    role="button"
    tabindex="0"
  >
    food
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-entertainment"
    role="button"
    tabindex="0"
  >
    entertainment
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-summer"
    role="button"
    tabindex="0"
  >
    summer
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-fall"
    role="button"
    tabindex="0"
  >
    fall
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-winter"
    role="button"
    tabindex="0"
  >
    winter
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-spring"
    role="button"
    tabindex="0"
  >
    spring
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-nature"
    role="button"
    tabindex="0"
  >
    nature
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-scenery"
    role="button"
    tabindex="0"
  >
    scenery
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-sky"
    role="button"
    tabindex="0"
  >
    sky
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-animal"
    role="button"
    tabindex="0"
  >
    animal
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-art"
    role="button"
    tabindex="0"
  >
    art
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-architecture"
    role="button"
    tabindex="0"
  >
    architecture
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-unique"
    role="button"
    tabindex="0"
  >
    unique
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-free"
    role="button"
    tabindex="0"
  >
    free
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-cheap"
    role="button"
    tabindex="0"
  >
    cheap
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-expensive"
    role="button"
    tabindex="0"
  >
    expensive
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-underground"
    role="button"
    tabindex="0"
  >
    underground
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-seasonal"
    role="button"
    tabindex="0"
  >
    seasonal
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-dangerous"
    role="button"
    tabindex="0"
  >
    dangerous
  </label>,
  <label
    class="tag btn btn-outline-primary"
    data-testid="tags"
    for="tgb-btn-other"
    role="button"
    tabindex="0"
  >
    other
  </label>,
]
`;

exports[`TagButtonGroup Component check to make sure provided tags are selected 1`] = `
<input
  autocomplete="off"
  checked=""
  class="btn-check"
  id="tgb-btn-river"
  type="checkbox"
  value="river"
/>
`;

exports[`TagButtonGroup Component check to make sure provided tags are selected 2`] = `
<input
  autocomplete="off"
  checked=""
  class="btn-check"
  id="tgb-btn-running"
  type="checkbox"
  value="running"
/>
`;

exports[`TagButtonGroup Component check to make sure provided tags are selected 3`] = `
<input
  autocomplete="off"
  checked=""
  class="btn-check"
  id="tgb-btn-fall"
  type="checkbox"
  value="fall"
/>
`;

exports[`TagButtonGroup Component check to make sure provided tags are selected 4`] = `
<input
  autocomplete="off"
  checked=""
  class="btn-check"
  id="tgb-btn-scenery"
  type="checkbox"
  value="scenery"
/>
`;

exports[`TagButtonGroup Component check to make sure provided tags are selected 5`] = `
<input
  autocomplete="off"
  checked=""
  class="btn-check"
  id="tgb-btn-seasonal"
  type="checkbox"
  value="seasonal"
/>
`;
